ndarray/src/main/java/org/tensorflow/ndarray/impl/sparse/AbstractSparseNdArray.java

```diff
@@ -0,0 +1,481 @@
+   *     example, {@code indices=[[1,3], [2,4]]} specifies that the elements with indexes of {@code
+   *     [1,3]} and {@code [2,4]} have nonzero values.
+   * @param values A 1-D NdArray of any type and shape {@code [N]}, which supplies the values for
+   *     each element in indices. For example, given {@code indices=[[1,3], [2,4]]}, the parameter
+   *     {@code values=[18, 3.6]} specifies that element {@code [1,3]} of the sparse NdArray has a
+   *     value of {@code 18}, and element {@code [2,4]} of the NdArray has a value of {@code 3.6}.
+   * @param dimensions the dimensional space for the dense object represented by this sparse array.
+   */
+  protected AbstractSparseNdArray(LongNdArray indices, U values, DimensionalSpace dimensions) {
+    super(dimensions);
+    this.indices = indices;
+    this.values = values;
+
+    // sanity checks on shapes, indices (shape = {@code [N, ndims]}, where N is the number of values
+    // (shape = {@code [N]}}.
+    if (this.indices.shape().get(0) != this.values.shape().get(0)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "The number of rows in indices (%d) does not  match the number of elements in values(%d).",
+              this.indices.shape().get(0), this.values.shape().get(0)));
+    }
+
+    // sanity checks on shapes, indices (shape = {@code [N, ndims]}, where ndims = the number of
+    // dimensions in the dense shape.
+    if (this.indices.shape().get(1) != shape().numDimensions()) {
+      throw new IllegalArgumentException(
+          String.format(
+              "The number of columns in indices (%d) does not  match the number of dimensions in shape (%d).",
+              this.indices.shape().get(1), shape().get(0)));
+    }
+  }
+
+  /**
+   * Creates an abstract SparseNdArray
+   *
+   * @param dimensions the dimensional space for the dense object represented by this sparse array,
+   */
+  protected AbstractSparseNdArray(DimensionalSpace dimensions) {
+    super(dimensions);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public NdArraySequence<U> elements(int dimensionIdx) {
+    if (dimensionIdx >= shape().numDimensions()) {
+      throw new IllegalArgumentException(
+          "Cannot iterate elements in dimension '"
+              + dimensionIdx
+              + "' of array with shape "
+              + shape());
+    }
+    if (rank() == 0 && dimensionIdx < 0) {
+      return new SingleElementSequence<>(this);
+    }
+    DimensionalSpace elemDims = dimensions().from(dimensionIdx + 1);
+
+    return new SlicingElementSequence<>(this, dimensionIdx, elemDims);
+  }
+
+  /**
+   * Computes the coordinates based on a relative position to the beginning of the dimension space.
+   *
+   * @param dimensions the dimension space
+   * @param position relative position to the beginning of the dimension space.
+   * @return the coordinates
+   */
+  protected long[] toCoordinates(DimensionalSpace dimensions, long position) {
```

Comment:
We should leave it like that for now but I think manipulating a `position` in a sparse array is wrong and it should have automatically access to the coordinates from which this position is coming from. But that will require some refactoring even at the dense level. 

Would you mind adding a "TODO" notice about that right here?

Reply:
OK